### PR TITLE
AWS IPAM Multi-Org ResourceDiscovery & ResourceDiscoveryAssociation

### LINKED ISSUE #29116: [New]: IPAM Resource Discovery & Association
### Description

New resources

`aws_vpc_ipam_resource_discovery`
`aws_vpc_ipam_resource_discovery_association`

### Requested Resource(s) and/or Data Source(s)

New resources

`aws_vpc_ipam_resource_discovery`
`aws_vpc_ipam_resource_discovery_association`

### Potential Terraform Configuration

_No response_

### References

https://aws.amazon.com/about-aws/whats-new/2023/01/amazon-vpc-ipam-ip-addresses-aws-organization/

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by drewmullen**
### Description

Add new IPAM Multi-organizational resource types and features:
- new resource:  resource discovery 
- new resource:  resource discovery association
- new attributes: ipam `default_resource_discovery_id` && `default_resource_discovery_association_id`
- new sweeper: add sweeper that removes all non-default resource discoveries and associations

### Relations

Closes #29116

### References

- [announcement](https://aws.amazon.com/about-aws/whats-new/2023/01/amazon-vpc-ipam-ip-addresses-aws-organization/)

### Output from Acceptance Testing

```
$ make testacc TESTS=TestAccIPAMResourceDiscovery_ PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccIPAMResourceDiscovery_'  -timeout 180m
=== RUN   TestAccIPAMResourceDiscovery_basic
=== PAUSE TestAccIPAMResourceDiscovery_basic
=== RUN   TestAccIPAMResourceDiscovery_modify
=== PAUSE TestAccIPAMResourceDiscovery_modify
=== RUN   TestAccIPAMResourceDiscovery_disappears
=== PAUSE TestAccIPAMResourceDiscovery_disappears
=== RUN   TestAccIPAMResourceDiscovery_tags
=== PAUSE TestAccIPAMResourceDiscovery_tags
=== CONT  TestAccIPAMResourceDiscovery_basic
=== CONT  TestAccIPAMResourceDiscovery_disappears
=== CONT  TestAccIPAMResourceDiscovery_modify
=== CONT  TestAccIPAMResourceDiscovery_tags
--- PASS: TestAccIPAMResourceDiscovery_disappears (24.35s)
--- PASS: TestAccIPAMResourceDiscovery_basic (26.70s)
--- PASS: TestAccIPAMResourceDiscovery_tags (44.87s)
--- PASS: TestAccIPAMResourceDiscovery_modify (84.00s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        84.112s

$ make testacc TESTS=TestAccIPAMResourceDiscoveryAssociation_ PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccIPAMResourceDiscoveryAssociation_'  -timeout 180m
=== RUN   TestAccIPAMResourceDiscoveryAssociation_basic
=== PAUSE TestAccIPAMResourceDiscoveryAssociation_basic
=== RUN   TestAccIPAMResourceDiscoveryAssociation_tags
=== PAUSE TestAccIPAMResourceDiscoveryAssociation_tags
=== RUN   TestAccIPAMResourceDiscoveryAssociation_disappears
=== PAUSE TestAccIPAMResourceDiscoveryAssociation_disappears
=== CONT  TestAccIPAMResourceDiscoveryAssociation_basic
=== CONT  TestAccIPAMResourceDiscoveryAssociation_disappears
=== CONT  TestAccIPAMResourceDiscoveryAssociation_tags
--- PASS: TestAccIPAMResourceDiscoveryAssociation_basic (37.96s)
--- PASS: TestAccIPAMResourceDiscoveryAssociation_disappears (38.71s)
--- PASS: TestAccIPAMResourceDiscoveryAssociation_tags (57.34s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        57.446s

$ make testacc TESTS=TestAccIPAM_ PKG=ec2                   
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccIPAM_'  -timeout 180m
=== RUN   TestAccIPAM_byoipIPv6
    ipam_byoip_test.go:22: Environment variable IPAM_BYOIP_IPV6_MESSAGE, IPAM_BYOIP_IPV6_SIGNATURE, or IPAM_BYOIP_IPV6_PROVISIONED_CIDR is not set
--- SKIP: TestAccIPAM_byoipIPv6 (0.00s)
=== RUN   TestAccIPAM_basic
=== PAUSE TestAccIPAM_basic
=== RUN   TestAccIPAM_disappears
=== PAUSE TestAccIPAM_disappears
=== RUN   TestAccIPAM_description
=== PAUSE TestAccIPAM_description
=== RUN   TestAccIPAM_operatingRegions
=== PAUSE TestAccIPAM_operatingRegions
=== RUN   TestAccIPAM_cascade
=== PAUSE TestAccIPAM_cascade
=== RUN   TestAccIPAM_tags
=== PAUSE TestAccIPAM_tags
=== CONT  TestAccIPAM_basic
=== CONT  TestAccIPAM_operatingRegions
=== CONT  TestAccIPAM_tags
=== CONT  TestAccIPAM_description
=== CONT  TestAccIPAM_cascade
=== CONT  TestAccIPAM_disappears
--- PASS: TestAccIPAM_disappears (26.89s)
--- PASS: TestAccIPAM_description (45.25s)
--- PASS: TestAccIPAM_tags (49.09s)
--- PASS: TestAccIPAM_basic (52.53s)
--- PASS: TestAccIPAM_cascade (73.24s)
--- PASS: TestAccIPAM_operatingRegions (85.92s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        86.041s

```

### CODE REVIEW

**Comment by drewmullen** [comment]
sweepers working:
1.  build a resource discovery
2. attach to an ipam
3. build another resource discovery
4. run sweepers

youll end up with no associations, and only the default resource discovery:
```
$ SWEEPARGS=-sweep-run=aws_vpc_ipam_resource_discovery make sweep
# make sweep SWEEPARGS=-sweep-run=aws_example_thing
# set SWEEPARGS=-sweep-allow-failures to continue after first failure
WARNING: This will destroy infrastructure. Use only in development accounts.
go test ./internal/sweep -v -tags=sweep -sweep=us-west-2,us-east-1,us-east-2 -sweep-run=aws_vpc_ipam_resource_discovery -timeout 60m
2023/02/02 14:31:26 [DEBUG] Running Sweepers for region (us-west-2):
2023/02/02 14:31:26 [DEBUG] Running Sweeper (aws_vpc_ipam_resource_discovery) in region (us-west-2)
2023/02/02 14:31:26 [INFO] Retrieved credentials from "EnvConfigCredentials"
2023/02/02 14:31:26 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023/02/02 14:31:27 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023/02/02 14:31:27 [DEBUG] Completed Sweeper (aws_vpc_ipam_resource_discovery) in region (us-west-2) in 672.158886ms
2023/02/02 14:31:27 Completed Sweepers for region (us-west-2) in 672.218998ms
2023/02/02 14:31:27 Sweeper Tests for region (us-west-2) ran successfully:
        - aws_vpc_ipam_resource_discovery
2023/02/02 14:31:27 [DEBUG] Running Sweepers for region (us-east-1):
2023/02/02 14:31:27 [DEBUG] Running Sweeper (aws_vpc_ipam_resource_discovery) in region (us-east-1)
2023/02/02 14:31:27 [INFO] Retrieved credentials from "EnvConfigCredentials"
2023/02/02 14:31:27 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023/02/02 14:31:27 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023/02/02 14:31:27 [DEBUG] Waiting for state to become: [success]
2023/02/02 14:31:27 [DEBUG] Waiting for state to become: [success]
2023/02/02 14:31:27 [DEBUG] Deleting IPAMResourceDiscovery: ipam-res-disco-009445ae4ba7118d9
2023/02/02 14:31:27 [DEBUG] Deleting IPAMResourceDiscovery: ipam-res-disco-0b6ee7c578c4683ae
2023/02/02 14:31:27 [DEBUG] Waiting for state to become: []
2023/02/02 14:31:27 [DEBUG] Waiting for state to become: []
2023/02/02 14:31:32 [DEBUG] Completed Sweeper (aws_vpc_ipam_resource_discovery) in region (us-east-1) in 5.496641656s
2023/02/02 14:31:32 Completed Sweepers for region (us-east-1) in 5.496662633s
2023/02/02 14:31:32 Sweeper Tests for region (us-east-1) ran successfully:
        - aws_vpc_ipam_resource_discovery
2023/02/02 14:31:32 [DEBUG] Running Sweepers for region (us-east-2):
2023/02/02 14:31:32 [DEBUG] Running Sweeper (aws_vpc_ipam_resource_discovery) in region (us-east-2)
2023/02/02 14:31:32 [INFO] Retrieved credentials from "EnvConfigCredentials"
2023/02/02 14:31:32 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023/02/02 14:31:33 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2023/02/02 14:31:33 [DEBUG] Completed Sweeper (aws_vpc_ipam_resource_discovery) in region (us-east-2) in 221.781903ms
2023/02/02 14:31:33 Completed Sweepers for region (us-east-2) in 221.805348ms
2023/02/02 14:31:33 Sweeper Tests for region (us-east-2) ran successfully:
        - aws_vpc_ipam_resource_discovery
ok      github.com/hashicorp/terraform-provider-aws/internal/sweep      6.414s
```

**Comment by drewmullen** [comment]
tests still passing after adding missed err check
```
$ make testacc TESTS=TestAccIPAMResourceDiscovery_ PKG=ec2       
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccIPAMResourceDiscovery_'  -timeout 180m
=== RUN   TestAccIPAMResourceDiscovery_basic
=== PAUSE TestAccIPAMResourceDiscovery_basic
=== RUN   TestAccIPAMResourceDiscovery_modify
=== PAUSE TestAccIPAMResourceDiscovery_modify
=== RUN   TestAccIPAMResourceDiscovery_disappears
=== PAUSE TestAccIPAMResourceDiscovery_disappears
=== RUN   TestAccIPAMResourceDiscovery_tags
=== PAUSE TestAccIPAMResourceDiscovery_tags
=== CONT  TestAccIPAMResourceDiscovery_basic
=== CONT  TestAccIPAMResourceDiscovery_disappears
=== CONT  TestAccIPAMResourceDiscovery_tags
=== CONT  TestAccIPAMResourceDiscovery_modify
--- PASS: TestAccIPAMResourceDiscovery_disappears (24.52s)
--- PASS: TestAccIPAMResourceDiscovery_basic (27.07s)
--- PASS: TestAccIPAMResourceDiscovery_tags (45.28s)
--- PASS: TestAccIPAMResourceDiscovery_modify (84.15s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        84.259s
```

File: internal/service/ec2/ipam_resource_discovery.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+			if len(operatingRegionUpdateAdd) != 0 {
+				input.AddOperatingRegions = operatingRegionUpdateAdd
+			}
+
+			if len(operatingRegionUpdateRemove) != 0 {
+				input.RemoveOperatingRegions = operatingRegionUpdateRemove
+			}
+		}
+
+		_, err := conn.ModifyIpamResourceDiscovery(input)
```
Comment (AdamTylerLynch) [inline internal/service/ec2/ipam_resource_discovery.go:201]: Please use the WithContext() method from the SDK.

File: internal/service/ec2/ipam_resource_discovery_association.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+	input := &ec2.AssociateIpamResourceDiscoveryInput{
+		ClientToken:             aws.String(resource.UniqueId()),
+		TagSpecifications:       tagSpecificationsFromKeyValueTags(tags, "ipam-resource-discovery-association"),
+		IpamId:                  aws.String(d.Get("ipam_id").(string)),
+		IpamResourceDiscoveryId: aws.String(d.Get("ipam_resource_discovery_id").(string)),
+	}
+
+	log.Printf("[DEBUG] Creating IPAM Resource Discovery Association: %s", input)
+	output, err := conn.AssociateIpamResourceDiscovery(input)
```
Comment (AdamTylerLynch) [inline internal/service/ec2/ipam_resource_discovery_association.go:94]: Please use WithContext() method from the SDK.

File: internal/service/ec2/ipam_resource_discovery_association.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+func ResourceIPAMResourceDiscoveryAssociationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	conn := meta.(*conns.AWSClient).EC2Conn()
+	defaultTagsConfig := meta.(*conns.AWSClient).DefaultTagsConfig
+	tags := defaultTagsConfig.MergeTags(tftags.New(d.Get("tags").(map[string]interface{})))
+
+	input := &ec2.AssociateIpamResourceDiscoveryInput{
+		ClientToken:             aws.String(resource.UniqueId()),
+		TagSpecifications:       tagSpecificationsFromKeyValueTags(tags, "ipam-resource-discovery-association"),
```
Comment (AdamTylerLynch) [inline internal/service/ec2/ipam_resource_discovery_association.go:88]: Can we pull this from the SDK as to not use a magic string?

File: internal/service/ec2/ipam_resource_discovery_association.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+func ResourceIPAMResourceDiscoveryAssociationDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	conn := meta.(*conns.AWSClient).EC2Conn()
+
+	input := &ec2.DisassociateIpamResourceDiscoveryInput{
+		IpamResourceDiscoveryAssociationId: aws.String(d.Id()),
+	}
+
+	log.Printf("[DEBUG] Disassociating IPAM Resource Discovery: %s", d.Id())
+	_, err := conn.DisassociateIpamResourceDiscovery(input)
```
Comment (AdamTylerLynch) [inline internal/service/ec2/ipam_resource_discovery_association.go:170]: Please use WithContext() method from the SDK.

File: website/docs/r/vpc_ipam_resource_discovery_association.html.markdown [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+* `is_default` - A boolean to identify if the Resource Discovery is the accounts default resource discovery
+* `owner_id` - The account ID for the account that manages the Resource Discovery
+* `ipam_resource_discovery_region` - The home region of the Resource Discovery
+* `ipam_region` - The home region of the IPAM
+* `ipam_arn` - The arn of the IPAM
+* `tags_all` - A map of tags assigned to the resource, including those inherited from the provider [`default_tags` configuration block](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#default_tags-configuration-block).
+
+## Import
+
+IPAMs can be imported using the `ipam resource discovery id`, e.g.
```
Comment (AdamTylerLynch) [inline website/docs/r/vpc_ipam_resource_discovery_association.html.markdown:53]: Should be association ID?

File: website/docs/r/vpc_ipam_resource_discovery_association.html.markdown [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+### operating_regions
+
+* `region_name` - (Required) The name of the Region you want to add to the IPAM.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - Amazon Resource Name (ARN) of IPAM Resource Discovery
+* `id` - The ID of the IPAM Resource Discovery
```
Comment (AdamTylerLynch) [inline website/docs/r/vpc_ipam_resource_discovery_association.html.markdown:43]: Should be association ID in the description?

File: website/docs/r/vpc_ipam_resource_discovery_association.html.markdown [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  ipam_id                    = aws_vpc_ipam.test.id
+  ipam_resource_discovery_id = aws_vpc_ipam_resource_discovery.test.id
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `description` - (Optional) A description for the IPAM Resource Discovery.
```
Comment (AdamTylerLynch) [inline website/docs/r/vpc_ipam_resource_discovery_association.html.markdown:30]:
```suggestion
* `description` - (Optional) A description for the IPAM Resource Discovery Association.
```

File: website/docs/r/vpc_ipam_resource_discovery_association.html.markdown [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+### operating_regions
+
+* `region_name` - (Required) The name of the Region you want to add to the IPAM.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `arn` - Amazon Resource Name (ARN) of IPAM Resource Discovery
```
Comment (AdamTylerLynch) [inline website/docs/r/vpc_ipam_resource_discovery_association.html.markdown:42]:
```suggestion
* `arn` - Amazon Resource Name (ARN) of IPAM Resource Discovery Association
```

**Review by AdamTylerLynch — CHANGES_REQUESTED** [review]
Requesting a few changes related to context passing and documentation. Thank you so much for this PR!

**Comment by AdamTylerLynch** [comment]
LGTM 🚀

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccIPAMResourceDiscovery_serial/ResourceDiscovery' PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20  -run=TestAccIPAMResourceDiscovery_serial/ResourceDiscovery -timeout 180m
=== RUN   TestAccIPAMResourceDiscovery_serial
=== PAUSE TestAccIPAMResourceDiscovery_serial
=== CONT  TestAccIPAMResourceDiscovery_serial
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscovery
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/basic
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/modify
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/disappears
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/tags
--- PASS: TestAccIPAMResourceDiscovery_serial (220.44s)
    --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscovery (220.44s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/basic (31.85s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/modify (101.10s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/disappears (26.86s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscovery/tags (60.62s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	228.465s
% make testacc TESTARGS='-run=TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation' PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20  -run=TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation -timeout 180m
=== RUN   TestAccIPAMResourceDiscovery_serial
=== PAUSE TestAccIPAMResourceDiscovery_serial
=== CONT  TestAccIPAMResourceDiscovery_serial
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation/basic
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation/disappears
=== RUN   TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation/tags
--- PASS: TestAccIPAMResourceDiscovery_serial (97.72s)
    --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation (147.72s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation/basic (39.28s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation/disappears (39.15s)
        --- PASS: TestAccIPAMResourceDiscovery_serial/ResourceDiscoveryAssociation/tags (63.53s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	154.254s
```

**Comment by ewbankkit** [comment]
@drewmullen Thanks for the contribution :tada: :clap:.

**Pull request merged**